File: Applied Course/4.Problem Solving/5.Problems on Trees/6.Print Ancestors of a given node in Binary Tree.c
#include<stdio.h> 
#include<stdlib.h> 

struct node 
{ 
   int data; 
   struct node* left; 
   struct node* right; 
}; 
  
/* If target element is present in tree, then prints the ancestors 
   and returns 1, otherwise returns 0. */
int printAncestors(struct node *root, int target) 
{ 
  if(!root) 
     return 0; 
  if(root->data == target) 
     return 1; 
  
  /* If target is present in either left or right subtree of this node, 
     then print this node */
  if(printAncestors(root->left, target) || printAncestors(root->right, target) ) 
  { 
    printf("%d ", root->data);
    return 1; 
  } 
  return 0; 
} 
  
/* create a new node*/
struct node* newNode(int data) 
{ 
  struct node* temp = (struct node*)malloc(sizeof(struct node)); 
  temp->data = data; 
  temp->left = NULL; 
  temp->right = NULL; 
  
  return temp; 
} 
  

int main() 
{ 
  struct node *root = newNode(10); 
  root->left        = newNode(20); 
  root->right       = newNode(30); 
  root->left->left  = newNode(40); 
  root->left->right = newNode(50); 
  root->left->left->left  = newNode(70); 
  
  printAncestors(root, 70); 
  
  getchar(); 
  return 0; 
} 

File: Applied Course/4.Problem Solving/5.Problems on Trees/6.Print Ancestors of a given node in Binary Tree.py
import sys
class node: 

    def __init__(self, info): 
        self.info = info  
        self.left = None
        self.right = None 
    
def insert(ptr,key):
    if(ptr is None):
        ptr=node(key)
    elif(key<=ptr.info):
        ptr.left=insert(ptr.left,key)
    elif(key>ptr.info):
        ptr.right=insert(ptr.right,key)
    return ptr

def printAncestors(root,target):
    if root is None:
        return False
    
    if root.info==target:
        return True
    
    if(printAncestors(root.left,target) or printAncestors(root.right,target)):
        print(root.info)
        return 1
    return 0

if __name__=='__main__':
    root=None
    root=insert(root,10)
    root=insert(root,5)
    root=insert(root,15)
    root=insert(root,30)
    printAncestors(root,30)
